src/main/java/seedu/todo/ui/InputHandler.java

```diff
@@ -1,12 +1,95 @@
 package seedu.todo.ui;
 
+import java.util.LinkedList;
+
 import seedu.todo.controllers.*;
 
 public class InputHandler {
+    
+    private static InputHandler instance;
+    
+    private static final int MAX_HISTORY_SIZE = 20;
+    private static final int INITIAL_HISTORY_INDEX = -1;
+    private static LinkedList<String> commandHistory = new LinkedList<String>();
+    private static int commandHistoryCurrentIndex = INITIAL_HISTORY_INDEX;
+
+    private Controller handlingController = null;
+    
+    protected InputHandler() {
+        // Prevent instantiation.
+    }
+    
+    /**
+     * Gets the current input handler instance.
+     */
+    public static InputHandler getInstance() {
+        if (instance == null) {
+            instance = new InputHandler();
+        }
+        
+        return instance;
+    }
+    
+    /**
+     * Pushes a command to the start of a LinkedList.
+     * @param command   Command string
+     */
+    private void pushCommand(String command) {
+        // Resets current pointer.
+        commandHistoryCurrentIndex = INITIAL_HISTORY_INDEX;
+        
+        // Adds to the start of the LinkedList.
+        commandHistory.addFirst(command);
+        
+        if (commandHistory.size() > MAX_HISTORY_SIZE) {
+            commandHistory.removeLast();
+        }
+    }
+    
+    /**
+     * Gets the previous command from the command history. Successive calls will return commands earlier in history.
+     * 
+     * @return  The input command earlier than what was previously retrieved
+     */
+    public String getPreviousCommandFromHistory() {
+        // Advance index only up to commandHistory.size()
+        // Advances down the list
+        if (commandHistoryCurrentIndex < commandHistory.size()) {
+            commandHistoryCurrentIndex++;
+        }
+        
+        // Returns empty String for index out of bounds
+        if (commandHistoryCurrentIndex >= commandHistory.size()) {
+            return "";
+        } else {
+            return commandHistory.get(commandHistoryCurrentIndex);
```

Comment:
Eh, I just realized this thing is O(n) too! Retrieving index in LinkedList is O(n), and turning this into a Deque would not work.

Your structure only needs to support small stuff (and user rarely needs to up/down) so I think your current LinkedList is fine, but for a proper O(1) you should implement two deques OR use .next() and .previous().